cmd/build/v2/build.go

```diff
@@ -218,125 +270,80 @@ func (ob *OktetoBuilder) Build(ctx context.Context, options *types.BuildOptions)
-
-			meta.Name = svcToBuild
-			meta.Namespace = ob.oktetoContext.GetNamespace()
-			meta.DevenvName = options.Manifest.Name
-			meta.RepoURL = ob.Config.GetAnonymizedRepo()
+	bg := newBuildGraph(buildManifest, toBuildSvcs)
+	tree, err := bg.GetGraph()
+	if err != nil {
+		return err
+	}
+	tree, err = tree.Subtree(toBuildSvcs...)
+	if err != nil {
+		return err
+	}
+	toBuildSvcs = tree.Ordered()
 
-			repoHashDurationStart := time.Now()
+	ob.ioCtrl.Logger().Infof("Images to build: [%s]", strings.Join(toBuildSvcs, ", "))
 
-			ob.ioCtrl.Logger().Debugf("getting project hash for analytics")
-			repoHash, err := ob.smartBuildCtrl.GetProjectHash(buildSvcInfo)
-			if err != nil {
-				ob.ioCtrl.Logger().Infof("error getting project commit hash: %s", err)
-			}
-			meta.RepoHash = repoHash
-			meta.RepoHashDuration = time.Since(repoHashDurationStart)
-
-			buildContextHashDurationStart := time.Now()
-
-			serviceHash := ob.smartBuildCtrl.GetServiceHash(buildSvcInfo, svcToBuild)
-			meta.BuildContextHash = serviceHash
-			meta.BuildContextHashDuration = time.Since(buildContextHashDurationStart)
-
-			// We only check that the image is built in the global registry if the noCache option is not set
-			if !options.NoCache && ob.smartBuildCtrl.IsEnabled() {
-				imageChecker := getImageChecker(ob.Config, ob.Registry, ob.smartBuildCtrl, ob.ioCtrl.Logger())
-				cacheHitDurationStart := time.Now()
-
-				buildHash := ob.smartBuildCtrl.GetBuildHash(buildSvcInfo, svcToBuild)
-				imageCtrl := registry.NewImageCtrl(ob.oktetoContext)
-				imageWithDigest, isBuilt := imageChecker.checkIfBuildHashIsBuilt(buildSvcInfo.Image, ob.oktetoContext.GetNamespace(), ob.oktetoContext.GetRegistryURL(), options.Manifest.Name, svcToBuild, buildHash, imageCtrl)
-
-				meta.CacheHit = isBuilt
-				meta.CacheHitDuration = time.Since(cacheHitDurationStart)
-
-				if isBuilt {
-					imageWithDigest, err := ob.smartBuildCtrl.CloneGlobalImageToDev(imageWithDigest, buildSvcInfo.Image)
-					if err != nil {
-						return err
-					}
-
-					ob.SetServiceEnvVars(svcToBuild, imageWithDigest)
-					builtImagesControl[svcToBuild] = true
-					meta.Success = true
-					ob.ioCtrl.Out().Infof("Okteto Smart Builds is skipping build of '%s' because it's already built from cache.", svcToBuild)
-					continue
-				}
-			}
+	sp := ob.ioCtrl.Out().Spinner("Collecting metadata for the images")
+	sp.Start()
+	if err := ob.metadataCollector.CollectMetadata(ctx, options.Manifest.Name, options.Manifest.Build, toBuildSvcs); err != nil {
+		return err
+	}
+	sp.Stop()
+	metaMap := ob.metadataCollector.GetMetadataMap()
+
+	notCachedServices := toBuildSvcs
+	var cachedServices []string
+	if !options.NoCache && ob.smartBuildCtrl.IsEnabled() {
+		sp := ob.ioCtrl.Out().Spinner("Checking if the images are already built from cache...")
+		sp.Start()
+		defer sp.Stop()
+		cachedServices, notCachedServices, err = ob.imageChecker.CheckImages(ctx, options.Manifest.Name, options.Manifest.Build, toBuildSvcs)
```

Comment:
Shouldn't this happen through the smart builds controller? I mean, at the end, smart builds is the cache enabler, but it sounds odd that the logic remains in other place. I think we should be coherent about it